File: 0001-two-sum/0001-two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        auto idx = vector<int>(nums.size());
        std::iota(idx.begin(), idx.end(), 0);
        std::sort(idx.begin(), idx.end(), [&nums](int l, int r) { return nums[l] < nums[r]; });
        int s = 0;
        int e = nums.size() - 1;
        
        while (s < e)
        {
            auto sum = nums[idx[s]] + nums[idx[e]];
            if (sum == target)
                break;
            if (sum > target)
                --e;
            else if (sum < target)
                ++s;
        }
        return vector<int>{idx[s], idx[e]};
    }
};

File: 0001-two-sum/0001-two-sum.cs
public class Solution {
    public int[] TwoSum(int[] nums, int target)
    {
        int[] indices = Enumerable.Range(0, nums.Length).ToArray();
        Array.Sort(indices, (l, r) => { return nums[l].CompareTo(nums[r]); });

        var l = 0;
        var r = nums.Length - 1;

        while (l < r)
        {
            var sum = nums[indices[l]] + nums[indices[r]];
            if (sum == target) return new[] { indices[l], indices[r] };
            if (sum > target) --r;
            else ++l;
        }

        return new[] { -1, -1 };
    }
}